File: src/application/useCases/importStatement.ts
import { hashContent, importFingerprint } from '../../domain/dedupe'
import type { Id, ImportBatch, SourceRecord, Transaction } from '../../domain/entities/types'
import { runPreview } from './importPreview'
import type {
  ImportPreview,
  ImportPreviewLine,
  ImportRequest,
  ImportStatementDeps,
} from './importTypes'

export type {
  ImportPreview,
  ImportPreviewLine,
  ImportRequest,
  ImportStatementDeps,
} from './importTypes'

/**
 * ImportStatement — قراءة الكشف، منع التكرار، إنشاء الدفعة.
 *
 * التدفق مرحلتان صراحةً (spec/04):
 *   preview()  → استخراج + مراجعة + ملخص الأثر، **بلا كتابة** (importPreview.ts)
 *   commit()   → تأكيد ذري داخل UnitOfWork
 *
 * السبب في الفصل: «إعادة الضغط أو إعادة فتح التطبيق لا تعيد الالتزام بالدفعة»
 * و«صفر أو كامل الدفعة، لا نصفها» (spec/05 و spec/06).
 */
export function makeImportStatement(deps: ImportStatementDeps) {
  async function preview(request: ImportRequest): Promise<ImportPreview> {
    return runPreview(deps, request)
  }

  function buildTransaction(
    id: Id,
    line: ImportPreviewLine,
    now: string,
    request: ImportRequest,
  ): Transaction {
    const txn: Transaction = {
      id,
      occurredAt: line.row.date,
      datePrecision: 'day',
      sourceOrder: line.row.lineNumber,
      // ⚠️ النوع الاقتصادي **لا يُستنتج من اتجاه السيولة** (spec/02).
      // يبدأ غير محدد وينتظر تأكيد المستخدم في المراجعة.
      economicKind: 'unclassified',
      economicKindConfirmed: false,
      observedDirection: line.row.direction,
      amountMinor: line.row.amountMinor,
      currency: 'SAR',
      categoryConfirmed: false,
      excludedFromBudget: false,
      reviewState: line.categoryId ? 'suggested' : 'needs_review',
      isCashTagged: false,
      rawDescription: line.row.description,
      rawMerchantName: line.row.merchantName,
      createdAt: now,
      updatedAt: now,
    }
    if (line.categoryId) txn.categoryId = line.categoryId
    // الرصيد المعلن يُحفظ كما ورد — هو مرجع المطابقة (OVERRIDES §7-ب)
    if (line.row.statedBalanceMinor !== undefined) {
      txn.statedBalanceMinor = line.row.statedBalanceMinor
    }
    if (request.walletId) txn.walletId = request.walletId
    if (line.row.sourceCategory) txn.sourceCategory = line.row.sourceCategory
    if (line.row.sourceOperationType) txn.sourceOperationType = line.row.sourceOperationType
    return txn
  }

  /**
   * المرحلة الثانية: الالتزام الذري.
   * `selectedLineNumbers` تسمح بتضمين متشابه أو استبعاد جديد **بقرار صريح**.
   */
  async function commitPrepared(
    request: ImportRequest,
    previewResult: ImportPreview,
    selectedLineNumbers?: readonly number[],
  ): Promise<ImportBatch> {
    const selection = selectedLineNumbers
      ? new Set(selectedLineNumbers)
      : new Set(previewResult.lines.filter((l) => l.selectedByDefault).map((l) => l.row.lineNumber))

    // التعارض لا يُكتب أبدًا بلا حسم صريح — لا استبدال صامت (spec/05)
    for (const line of previewResult.lines) {
      if (selection.has(line.row.lineNumber) && line.state === 'conflict') {
        throw new Error(`الصف ${line.row.lineNumber} فيه تعارض ولازم تحسمه الأول. ${line.reason}`)
      }
    }

    return deps.uow.run(async () => {
      const batchId = deps.ids.next('batch')
      const now = deps.clock.nowIso()
      const transactions: Transaction[] = []
      const records: SourceRecord[] = []

      const counts = {
        total: previewResult.counts.total,
        imported: 0, // يُملأ بعد البناء
        duplicates: previewResult.counts.duplicates,
        similar: previewResult.counts.similar,
        conflicts: previewResult.counts.conflicts,
        invalid: previewResult.counts.invalid,
      }

      for (const line of previewResult.lines) {
        const included = selection.has(line.row.lineNumber)
        const txnId = included ? deps.ids.next('txn') : null
        if (txnId) transactions.push(buildTransaction(txnId, line, now, request))

        records.push({
          id: deps.ids.next('src'),
          batchId,
          accountIdentity: request.accountIdentity,
          sourceReference: line.row.reference,
          sourceHash: hashContent(line.row.raw),
          originalRowIndex: line.row.lineNumber,
          rawLine: line.row.raw,
          transactionId: txnId,
          matchingState: line.state,
          reason: line.reason,
        })
      }

      counts.imported = transactions.length

      /*
       * بروتوكول علامة الالتزام — يحقق «صفر أو كامل الدفعة» (spec/06)
       * حتى على Firestore حيث لا تغطي المعاملة الواحدة مئات المستندات.
       *
       *   ١. الدفعة تُكتب `staged` **قبل** أي شيء
       *   ٢. **سجلات المصدر أولًا** ثم العمليات (الانقطاع المحتمل هنا)
       *   ٣. التحويل إلى `committed` بكتابة واحدة ذرّية بطبيعتها
       *
       * ⚠️ **ترتيب الخطوة ٢ ليس تفصيلًا.** سجل المصدر هو الفهرس الوحيد
       * الذي يربط العملية بدفعتها. لو كُتبت العمليات أولًا وانقطع الاتصال
       * قبل السجلات، تصير العمليات **يتيمة**: موجودة في قاعدة البيانات
       * ولا سبيل للوصول إليها من الدفعة، فيستحيل تنظيفها.
       *
       * بهذا الترتيب أي انقطاع يترك سجلات تُشير إلى عمليات قد لا تكون
       * كُتبت بعد — وحذف معرّف غير موجود لا يضر — فالتنظيف يعمل دائمًا.
       *
       * أي انقطاع يترك دفعة `staged`، و findByFileHash لا يرى إلا
       * `committed`، فما تحتها غير معتمد ويُنظَّف بـ ResumeStagedBatch.
       */
      const batch: ImportBatch = {
        id: batchId,
        sourceType: request.sourceType,
        fileHash: previewResult.fileHash,
        fileName: request.fileName,
        importedAt: now,
        state: 'staged',
        counts,
      }

      await deps.batches.save(batch) // ١
      await deps.sources.saveMany(records) // ٢ — الفهرس أولًا
      await deps.txns.saveMany(transactions) //     ثم العمليات
      await deps.batches.updateState(batchId, 'committed') // ٣

      return { ...batch, state: 'committed' }
    })
  }


  let committing = false
  async function commit(request:ImportRequest, previous:ImportPreview, selected?:readonly number[]):Promise<ImportBatch> {
    if (committing) throw new Error('فيه استيراد بيتحفظ حاليًا؛ استنى اكتماله')
    committing = true
    try {
      if (importFingerprint(request.content, request.accountIdentity) !== previous.fileHash || request.accountIdentity !== previous.accountIdentity) {
        throw new Error('بيانات الاستيراد اتغيرت؛ اعمل معاينة جديدة')
      }
      const fresh = await preview(request)
      const selection = selected ?? previous.lines.filter(line=>line.selectedByDefault).map(line=>line.row.lineNumber)
      /*
       * الملف ده اتستورد قبل كده؟ نرجّع الدفعة القديمة **بس لو مفيش حاجة
       * جديدة تتضاف**. قرار المالك 2026-09-12: الناس بترفع ملفات كتير
       * متداخلة، والملف اللي اتستورد منه جزء لازم يكمل. منع التكرار بيشتغل
       * **صف بصف** أصلًا، فمفيش خطر تكرار من إكمال الملف.
       * (قبل كده: أي ملف بصمته متسجلة كان بيترفض بصمت — 1077 عملية حقيقية
       * ما دخلتش والورقة قفلت كأنها نجحت.)
       */
      const addable = selection.filter((number) => {
        const line = fresh.lines.find((l) => l.row.lineNumber === number)
        return line !== undefined && line.state !== 'duplicate' && line.state !== 'invalid'
      })
      if (fresh.previousBatch && addable.length === 0) return fresh.previousBatch
      for (const number of selection) {
        const before = previous.lines.find(line=>line.row.lineNumber===number)
        const now = fresh.lines.find(line=>line.row.lineNumber===number)
        if (!before || !now || JSON.stringify(before.row)!==JSON.stringify(now.row) || now.state!==before.state || now.matchedTransactionId!==before.matchedTransactionId) {
          throw new Error('فيه بيانات اتغيرت بعد المعاينة؛ اعمل معاينة جديدة علشان نمنع التكرار')
        }
        if (now.state === 'duplicate' || now.state === 'invalid') throw new Error('العملية المكررة أو غير الصالحة مش قابلة للإضافة')
      }
      return await commitPrepared(request, fresh, selection)
    } finally { committing = false }
  }

  return { preview, commit }
}

File: native-app/app/src/commonMain/kotlin/app/masroufy/usecase/ImportStatement.kt
package app.masroufy.usecase

import app.masroufy.core.Currency
import app.masroufy.core.EconomicKind
import app.masroufy.core.Id
import app.masroufy.core.ImportBatch
import app.masroufy.core.ImportBatchState
import app.masroufy.core.ImportCounts
import app.masroufy.core.MatchingState
import app.masroufy.core.ReviewState
import app.masroufy.core.SourceRecord
import app.masroufy.core.Transaction
import app.masroufy.core.hashContent
import app.masroufy.core.importFingerprint

/**
 * ImportStatement — نقل `importStatement.ts`: قراءة الكشف، منع التكرار، إنشاء الدفعة.
 * التدفق مرحلتين صراحةً (spec/04): `preview()` من غير كتابة، و`commit()` ذري جوه UnitOfWork.
 */
class ImportStatement(private val deps: ImportStatementDeps) {
    suspend fun preview(request: ImportRequest): ImportPreview = runPreview(deps, request)

    private fun buildTransaction(id: Id, line: ImportPreviewLine, now: String, request: ImportRequest, chosenCategoryId: Id?): Transaction {
        // تصنيف اختاره المستخدم بنفسه وقت المراجعة = مؤكد، زي `categorizeTransactions.confirm`
        val categoryId = chosenCategoryId ?: line.categoryId
        return Transaction(
            id = id,
            occurredAt = line.row.date,
            datePrecision = "day",
            sourceOrder = line.row.lineNumber,
            // ⚠️ النوع الاقتصادي **ما بيستنتجش من اتجاه السيولة** (spec/02) — بيبدأ غير محدد
            economicKind = EconomicKind.UNCLASSIFIED,
            economicKindConfirmed = false,
            observedDirection = line.row.direction,
            amountMinor = line.row.amountMinor,
            currency = Currency.SAR,
            categoryConfirmed = chosenCategoryId != null,
            excludedFromBudget = false,
            reviewState = when {
                chosenCategoryId != null -> ReviewState.CONFIRMED
                line.categoryId != null -> ReviewState.SUGGESTED
                else -> ReviewState.NEEDS_REVIEW
            },
            isCashTagged = false,
            rawDescription = line.row.description,
            rawMerchantName = line.row.merchantName,
            createdAt = now,
            updatedAt = now,
            categoryId = categoryId,
            // الرصيد المعلن بيتحفظ زي ما ورد — هو مرجع المطابقة (OVERRIDES §7-ب)
            statedBalanceMinor = line.row.statedBalanceMinor,
            walletId = request.walletId,
            sourceCategory = line.row.sourceCategory,
            sourceOperationType = line.row.sourceOperationType,
        )
    }

    /** المرحلة التانية: الالتزام الذري. `selectedLineNumbers` بتسمح بضم متشابه أو استبعاد جديد **بقرار صريح**. */
    private suspend fun commitPrepared(
        request: ImportRequest,
        previewResult: ImportPreview,
        selectedLineNumbers: List<Int>?,
        chosenCategories: Map<Int, Id>?,
    ): ImportBatch {
        val selection = selectedLineNumbers?.toSet()
            ?: previewResult.lines.filter { it.selectedByDefault }.map { it.row.lineNumber }.toSet()

        // التعارض ما بيتكتبش أبدًا من غير حسم صريح — لا استبدال صامت (spec/05)
        for (line in previewResult.lines) {
            if (line.row.lineNumber in selection && line.state == MatchingState.CONFLICT) {
                throw IllegalStateException("الصف ${line.row.lineNumber} فيه تعارض ولازم تحسمه الأول. ${line.reason}")
            }
        }

        return deps.uow.run {
            val batchId = deps.ids.next("batch")
            val now = deps.clock.nowIso()
            val transactions = mutableListOf<Transaction>()
            val records = mutableListOf<SourceRecord>()

            for (line in previewResult.lines) {
                val included = line.row.lineNumber in selection
                val txnId = if (included) deps.ids.next("txn") else null
                if (txnId != null) transactions += buildTransaction(txnId, line, now, request, chosenCategories?.get(line.row.lineNumber))

                records += SourceRecord(
                    id = deps.ids.next("src"),
                    batchId = batchId,
                    accountIdentity = request.accountIdentity,
                    sourceReference = line.row.reference,
                    sourceHash = hashContent(line.row.raw),
                    originalRowIndex = line.row.lineNumber,
                    rawLine = line.row.raw,
                    transactionId = txnId,
                    matchingState = line.state,
                    reason = line.reason,
                )
            }

            /*
             * بروتوكول علامة الالتزام — «صفر أو كامل الدفعة» (spec/06):
             *   ١. الدفعة بتتكتب `staged` قبل أي حاجة
             *   ٢. **سجلات المصدر الأول** وبعدها العمليات — السجل هو الفهرس الوحيد
             *      اللي بيربط العملية بدفعتها؛ لو العمليات اتكتبت الأول واتقطع
             *      الاتصال، بتبقى يتيمة ومفيش طريق لتنظيفها
             *   ٣. التحويل لـ`committed` بكتابة واحدة ذرّية بطبيعتها
             */
            val batch = ImportBatch(
                id = batchId,
                sourceType = request.sourceType,
                fileHash = previewResult.fileHash,
                fileName = request.fileName,
                importedAt = now,
                state = ImportBatchState.STAGED,
                counts = ImportCounts(
                    total = previewResult.counts.total,
                    imported = transactions.size,
                    duplicates = previewResult.counts.duplicates,
                    similar = previewResult.counts.similar,
                    conflicts = previewResult.counts.conflicts,
                    invalid = previewResult.counts.invalid,
                ),
            )

            deps.batches.save(batch) // ١
            deps.sources.saveMany(records) // ٢ — الفهرس الأول
            deps.txns.saveMany(transactions) //     وبعده العمليات
            deps.batches.updateState(batchId, ImportBatchState.COMMITTED) // ٣

            batch.copy(state = ImportBatchState.COMMITTED)
        }
    }

    private var committing = false

    /** `chosenCategories`: تصنيف اختاره المستخدم لسطر (رقم السطر ← التصنيف) — بيتحفظ مؤكد. */
    suspend fun commit(
        request: ImportRequest,
        previous: ImportPreview,
        selected: List<Int>? = null,
        chosenCategories: Map<Int, Id>? = null,
    ): ImportBatch {
        if (committing) throw IllegalStateException("فيه استيراد بيتحفظ حاليًا؛ استنى اكتماله")
        committing = true
        try {
            if (importFingerprint(request.content, request.accountIdentity) != previous.fileHash || request.accountIdentity != previous.accountIdentity) {
                throw IllegalStateException("بيانات الاستيراد اتغيرت؛ اعمل معاينة جديدة")
            }
            val fresh = preview(request)
            val selection = selected ?: previous.lines.filter { it.selectedByDefault }.map { it.row.lineNumber }
            /*
             * الملف اتستورد قبل كده؟ بنرجّع الدفعة القديمة **بس لو مفيش حاجة جديدة تتضاف**
             * (قرار المالك 2026-09-12: الملف اللي اتستورد منه جزء لازم يكمل — منع التكرار
             * بيشتغل صف بصف أصلًا).
             */
            val addable = selection.filter { number ->
                val line = fresh.lines.find { it.row.lineNumber == number }
                line != null && line.state != MatchingState.DUPLICATE && line.state != MatchingState.INVALID
            }
            if (fresh.previousBatch != null && addable.isEmpty()) return fresh.previousBatch
            for (number in selection) {
                val before = previous.lines.find { it.row.lineNumber == number }
                val now = fresh.lines.find { it.row.lineNumber == number }
                if (before == null || now == null || before.row != now.row || now.state != before.state || now.matchedTransactionId != before.matchedTransactionId) {
                    throw IllegalStateException("فيه بيانات اتغيرت بعد المعاينة؛ اعمل معاينة جديدة علشان نمنع التكرار")
                }
                if (now.state == MatchingState.DUPLICATE || now.state == MatchingState.INVALID) {
                    throw IllegalStateException("العملية المكررة أو غير الصالحة مش قابلة للإضافة")
                }
            }
            return commitPrepared(request, fresh, selection, chosenCategories)
        } finally {
            committing = false
        }
    }
}
